lib/psql.js

```diff
@@ -24,32 +16,115 @@ function env (db) {
   })
 }
 
+function tunnelConfig (db) {
+  const localHost = '127.0.0.1'
+  const localPort = Math.floor(Math.random() * (65535 - 49152) + 49152)
+  return {
+    username: 'bastion',
+    host: db.bastionHost,
+    privateKey: db.bastionKey,
+    dstHost: db.host,
+    dstPort: db.port,
+    localHost: localHost,
+    localPort: localPort
+  }
+}
+
+function handlePsqlError (reject, psql) {
+  psql.on('error', (err) => {
+    if (err.code === 'ENOENT') {
+      reject(`The local psql command could not be located. For help installing psql, see https://devcenter.heroku.com/articles/heroku-postgresql#local-setup`)
+    } else {
+      Promise.reject(err)
+    }
+  })
+}
+
+function sshTunnel (db, dbTunnelConfig, timeout) {
+  return new Promise((resolve, reject) => {
+    // if necessary to tunnel, setup a tunnel
+    // see also https://github.com/heroku/heroku/blob/master/lib/heroku/helpers/heroku_postgresql.rb#L53-L80
+    if (db.bastionKey) {
+      tunnel(dbTunnelConfig, (err, tnl) => {
+        if (err) {
+          debug(err)
+          reject(`Unable to establish a secure tunnel to your database.`)
+        }
+        debug('Tunnel created')
+        resolve(tnl)
+      })
+    } else {
+      resolve()
+    }
+    setTimeout(() => reject('Establishing a secure tunnel timed out'), timeout)
```

Comment:
Oh ugh I just realized that if you do this I think you need to cancel the timeout (with the token from the return value of `setTimeout`) when the tunnel does resolve successfully (or isn't needed). I think it's also fine to skip the timeout instead.